libs/sdk-ui-dashboard/src/model/queryServices/queryInsightDateDatasets.ts

```diff
@@ -0,0 +1,287 @@
+}
+
+/**
+ * Given insight and a list of available cataloged date datasets, this function looks up date datasets that are used
+ * in insight's global date filters. Date datasets
+ *
+ * @param insight - insight to lookup date dataset for
+ * @param datasets - all cataloged date datasets
+ */
+function lookupDatasetsUsedInDateFilters(
+    insight: IInsight,
+    datasets: ICatalogDateDataset[],
+): ReadonlyArray<ICatalogDateDataset> {
+    const dateFilters = insightFilters(insight).filter(isDateFilter);
+
+    return dateFilters.map((filter) => {
+        const datasetRef = filterObjRef(filter);
+        const dateDataset = datasets.find((dataset) => areObjRefsEqual(datasetRef, dataset.dataSet.ref));
+
+        // if this happens then either component is using wrong catalog or metadata is inconsistent (and insight thus un-renderable)
+        invariant(dateDataset, `cannot find metadata for date dataset ${objRefToString(datasetRef)}`);
+
+        return dateDataset;
+    });
+}
+
+/**
+ * Given insight and list of available cataloged date datasets, this generator will first query usage & metadata about
+ * used display forms and attributes. With this, the function will lookup date datasets for each display form
+ * used in insight's attribute buckets and attribute filters.
+ *
+ * Note: that cataloged date datasets already contain mapping of dataset -> attribute -> default display form. However
+ * this cannot be used because the code cannot expect that the insights only use date dataset's default display forms.
+ *
+ * @param insight - insight work with
+ * @param datasets - cataloged date datasets
+ */
+function* lookupDatasetsUsedInAttributesAndFilters(insight: IInsight, datasets: ICatalogDateDataset[]) {
+    const insightAttributes: InsightAttributesMeta = yield call(
+        query,
+        queryInsightAttributesMeta(insightRef(insight)),
+    );
+    const {
+        usage: { inAttributes, inFilters },
+        displayForms,
+    } = insightAttributes;
+
+    const datasetLookup: (ref: ObjRef) => ICatalogDateDataset | undefined = (ref) => {
+        const displayForm = displayForms[serializeObjRef(ref)];
+
+        // if this bombs then the query insight attributes is messed up as it does not include display forms meta
+        // for some display form used in the insight
+        invariant(displayForm);
+
+        const attributeRef = displayForm.attribute;
+        return datasets.find((dataset) => {
+            return dataset.dateAttributes.find((attribute) => {
+                return areObjRefsEqual(attribute.attribute.ref, attributeRef);
+            });
+        });
+    };
+
+    return {
+        usedInAttributes: inAttributes.map(datasetLookup),
+        usedInAttributeFilters: inFilters.map(datasetLookup),
+    };
+}
+
+function* lookupDatasetsInInsight(insight: IInsight) {
+    /*
+     * XXX: still contemplating here that looking up against catalog date datasets is unnecessary as
```

Comment:
Is this decidable now?

Reply:
nope not yet. i've changed this to TODO & added more docs / explanation. the way we have it now matches the previous logic in KD.